File: kt/src/main/kotlin/com/github/ephemient/aoc2025/Day8.kt
package com.github.ephemient.aoc2025

fun day8(input: String, n: Int = 1000): Pair<Int, Long?> {
    val nodes = input.lines().mapNotNull { line ->
        line.split(',').mapNotNull { it.toLongOrNull() }.ifEmpty { null }
    }
    val edges = buildList {
        for ((i, node) in nodes.withIndex()) {
            for (j in i + 1 ..< nodes.size) {
                add(Triple(i, j, node.zip(nodes[j], Long::minus).sumOf { it * it }))
            }
        }
        sortBy { it.third }
    }
    var components = nodes.size
    val mapping = mutableMapOf<Int, Int>()
    fun lookup(key: Int): Int {
        val value = mapping[key] ?: return key
        return lookup(value).also { mapping[key] = it }
    }
    var part1 = 0
    edges.forEachIndexed { index, (i, j) ->
        val a = lookup(i)
        val b = lookup(j)
        if (a != b) {
            mapping[maxOf(a, b)] = minOf(a, b)
            components--
        }
        if (index + 1 == n) {
            val sizes = nodes.indices.groupingBy(::lookup).fold(0) { acc, _ -> acc + 1 }
            if (sizes.size >= 3) part1 = sizes.values.sorted().takeLast(3).fold(1, Int::times)
        }
        if (components == 1) return part1 to nodes[i][0] * nodes[j][0]
    }
    return part1 to null
}

File: py/src/aoc2025/day8.py
from collections import defaultdict
from math import prod
from operator import itemgetter


def _day8(input: str, n: int = 1000) -> tuple[int, int | None]:
    nodes = [tuple(map(int, line.split(","))) for line in input.splitlines()]
    part1 = 0
    components = len(nodes)
    mapping: dict[int, int] = {}

    def lookup(key: int) -> int:
        if key not in mapping:
            return key
        value = lookup(mapping[key])
        mapping[key] = value
        return value

    for m, (i, j, _) in enumerate(
        sorted(
            (
                (i, j, sum((a - b) * (a - b) for a, b in zip(node, nodes[j])))
                for (i, node) in enumerate(nodes)
                for j in range(i + 1, len(nodes))
            ),
            key=itemgetter(2),
        )
    ):
        a, b = lookup(i), lookup(j)
        if a != b:
            mapping[max(a, b)] = min(a, b)
            components -= 1
        if m + 1 == n:
            groupings = defaultdict(set)
            for key in range(len(nodes)):
                groupings[lookup(key)].add(key)
            if len(groupings) >= 3:
                part1 = prod(sorted(map(len, groupings.values()))[-3:])
        if components == 1:
            return part1, nodes[i][0] * nodes[j][0]

    return part1, None


parts = (_day8,)
